### ISSUE TITLE
Add shortcut to trigger tab slide

### ISSUE THREAD

**Issue by nkakouros** [ZitRos/tab-slider#6]
I often want to quickly switch to a secondary tab from the main one I am currently working on in order to look at sth quickly and then go back to the main one to continue working. For this reason, I have set the slide delay to 15 seconds. But, some other times I want to go to some tab that lies quite on the right and make it the main one. I have to either drag and drop it to move it to the leftmost position or wait for 15 seconds for the addon to trigger. It would be nice to have a customizable shortcut to trigger tab slider and have the currently active tab slide at once.

**Comment by nikitaeverywhere** [comment]
Hello! Thanks for the suggestion. Here are a few thoughts on it:

- Did you know what if (say you have a default 1 sec delay) you left mouse key press on the tab for more than 1 second it won't slide even after you release?
- The ideal scenario, which is unfortunately limited by browser APIs now is: you have 1 second delay (or even no delay at all), and the tab slides ONLY after you release Ctrl click (when switching with keyboard) or mouse button. Unfortunately, it works only for mouse and there's no way to tell programatically in the extension whether Ctrl is pressed. The closest is `chrome.commands`, but it doesn't work to detect tab switching ([Tab was removed from list of supported keys in Chrome 33 for accessibility reasons.](https://developer.chrome.com/docs/extensions/reference/commands/) etc).
- I also don't want to inject a script to every page to watch for key bindings, as it will into more permissions on the extension. This is debatable (anyone is welcome to share the opinion - would you still be OK if this extension inject scripts?)
- Setting 15 seconds delay isn't how this supposed to work... Try this :) Personally I do peek tabs from time to time, but as I install this extension the "order of the tabs" doesn't matter anymore to me. If I want to keep something ordered/pinned, I use pinned tabs. It's okay for me to look quickly at the tab and for it to slide left - then I just Ctrl+Tab (maybe a few times) to the previous one where I was before.
- As a temporary workaround, I imagine what can be done is this: introduce, for example, "Ctrl+Q" and "Ctrl+Shift+Q" (or Ctrl+\` and Ctrl+Shift+\`) hotkeys for tab switching, that will enable "the ideal scenario" described above but will be used instead of Ctrl+Tab for switching tabs. What do you think?

**Comment by nkakouros** [comment]
Thank you for the explanation.

>     * Did you know what if (say you have a default 1 sec delay) you left mouse key press on the tab for more than 1 second it won't slide even after you release?

No, I will keep this in mind.

>     * I also don't want to inject a script to every page to watch for key bindings, as it will into more permissions on the extension. This is debatable (anyone is welcome to share the opinion - would you still be OK if this extension inject scripts?)

I don't have an issue as long as I trust the extension and it doesn't crash the website.

>     * Setting 15 seconds delay isn't how this supposed to work... Try this :) Personally I do peek tabs from time to time, but as I install this extension the "order of the tabs" doesn't matter anymore to me. If I want to keep something ordered/pinned, I use pinned tabs. It's okay for me to look quickly at the tab and for it to slide left - then I just Ctrl+Tab (maybe a few times) to the previous one where I was before.

There are some issues with using keyboard shortcuts in my setup. One is that on an apple keyboard Ctrl+Tab is an uncomfortable key combination. Then, changing built-in shortcuts is not supported by Firefox natively. One has to use addons that support this, but one that I found did not work for the key combinations that were natural for me. It could be that macos support was not good or that the shortcuts were consumed by the window manager/system instead of the application (firefox). But I also use Tridactyl to switch to tabs sometimes (with `<space>n`, `<space>p` for next and previous tab respectively). After your comment I discovered the `tabmove 0` command that does what I want. Theoretically, I should be able to create a custom shortcut in Tridactyl to trigger that command, but I again did not succeed in creating comfortable mappings. In any case, I think that the Tridactyl solution is good enough for me. I will consider your advice on how to use Tab Slider as well.

**Comment by nikitaeverywhere** [comment]
Sad to hear that the default key bindings aren't convenient for you. I'm pretty sure there's no "nice" way to fix it, rather than just stick to this combination. It works quite nicely for me and I believe most of users.

TL;DR the most likely next step I can take for TabSlider is inserting content script to all browser pages just to track CTRL press - the default behaviour would be to move tabs around only when CTRL/CMD is released. The immediate problem, however, is that it will work only when page is focused. Most of cases will be covered except when users start from DevTools, for instance, or switch tabs after pressing on out-of-the-page elements like extensions or navbar.

I'll keep this open for a while to maybe hear other opinions.

**Comment by nkakouros** [comment]
On a similar note, it would be convenient to have a toolbar button to toggle the extension's behavior on and off. This way, tab sliding could be set to happen instantaneously (as you mention that it is supposed to be used) but for sessions where the user needs an order in their tabs, they could quickly turn the sliding off.

**Comment by nikitaeverywhere** [comment]
> a toolbar button to toggle the extension's behavior on and off

Makes sense, however it sounds more like an exceptional behaviour for me. My line of thinking is: you either get comfortable with the sorting behaviour or give up. Mixing things from time to time would make you think the tab should have slided where it won't :)

**Comment by nkakouros** [comment]
> Did you know what if (say you have a default 1 sec delay) you left mouse key press on the tab for more than 1 second it won't slide even after you release?

Actually, I tried this on a fresh Firefox installation and it does not work, the tab just slides regardless of me releasing the mouse or not. I am on a Macbook using the touchpad if it makes any difference.

**Comment by nikitaeverywhere** [comment]
> Actually, I tried this on a fresh Firefox installation and it does not work, the tab just slides regardless of me releasing the mouse or not. I am on a Macbook using the touchpad if it makes any difference.

Indeed. It works in Chrome tho. It's not an intended behaviour tho (I don't use it), but might be useful.
